mlrun/launcher/base.py

```diff
@@ -49,11 +49,38 @@ def verify_base_image(runtime):
         """resolves and sets the build base image if build is needed"""
         pass
 
-    @staticmethod
-    @abc.abstractmethod
-    def save(runtime):
-        """store the function to the db"""
-        pass
+    def save_function(
+        self,
+        runtime: "mlrun.runtimes.BaseRuntime",
+        tag: str = "",
+        versioned: bool = False,
+        refresh: bool = False,
+    ) -> str:
+        """
+        store the function to the db
+        :param runtime:     runtime object
+        :param tag:         function tag to store
+        :param versioned:   whether we want to version this function object so that it will queryable by its hash key
+        :param refresh:     refresh function metadata
+
+        :return:            function uri
+        """
+        if not self.db:
+            logger.error("database connection is not configured")
+            return ""
```

Comment:
so, raise an exception? im not sure when it is valid and returning empty string is ok